File: expedfacil/src/main/java/com/example/expedfacil/bussiness/UsuarioService.java
package com.example.expedfacil.bussiness;

import java.time.LocalDateTime;

import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;

import com.example.expedfacil.bussiness.exception.ConflitoException;
import com.example.expedfacil.bussiness.exception.RecursoNaoEncontradoException;
import com.example.expedfacil.controller.dto.usuario.UsuarioCreateRequest;
import com.example.expedfacil.controller.dto.usuario.UsuarioResponse;
import com.example.expedfacil.controller.dto.usuario.UsuarioSenhaUpdateRequest;
import com.example.expedfacil.controller.dto.usuario.UsuarioUpdateRequest;
import com.example.expedfacil.infrastructure.entitys.Usuario;
import com.example.expedfacil.infrastructure.entitys.enums.RoleUsuario;
import com.example.expedfacil.infrastructure.repository.UsuarioRepository;

import lombok.RequiredArgsConstructor;

@Service
@RequiredArgsConstructor
public class UsuarioService {

    private final UsuarioRepository usuarioRepository;
    private final PasswordEncoder passwordEncoder;

    public Page<UsuarioResponse> listarAtivos(Pageable pageable) {
        return usuarioRepository.findByAtivoTrue(pageable).map(this::toResponse);
    }

    public void criarUsuario(UsuarioCreateRequest request) {
        if (usuarioRepository.existsByMatricula(request.matricula())) {
            throw new ConflitoException("Ja existe um usuario com matricula " + request.matricula() + ".");
        }

        LocalDateTime now = LocalDateTime.now();

        Usuario usuario = Usuario.builder()
                .matricula(request.matricula())
                .nome(request.nome().trim())
                .cargo(request.cargo())
                .senhaHash(passwordEncoder.encode(request.senha()))
                .role(RoleUsuario.USER)
                .ativo(true)
                .createdAt(now)
                .updatedAt(now)
                .build();

        usuarioRepository.save(usuario);
    }

    public void atualizarUsuario(Long id, UsuarioUpdateRequest request) {
        Usuario usuario = buscarAtivoPorId(id);

        usuario.setNome(request.nome().trim());
        usuario.setCargo(request.cargo());

        if (request.senha() != null && !request.senha().isBlank()) {
            usuario.setSenhaHash(passwordEncoder.encode(request.senha()));
        }

        usuario.setUpdatedAt(LocalDateTime.now());
        usuarioRepository.save(usuario);
    }

    public void atualizarSenha(Long id, UsuarioSenhaUpdateRequest request) {
        Usuario usuario = buscarAtivoPorId(id);
        usuario.setSenhaHash(passwordEncoder.encode(request.senha()));
        usuario.setUpdatedAt(LocalDateTime.now());
        usuarioRepository.save(usuario);
    }

    public void desativarUsuario(Long id, Integer matriculaLogada) {
        Usuario usuario = usuarioRepository.findById(id)
                .orElseThrow(() -> new RecursoNaoEncontradoException("Usuario nao encontrado para o id " + id));

        if (!usuario.isAtivo()) {
            throw new ConflitoException("Usuario ja esta inativo.");
        }

        if (usuario.getMatricula().equals(matriculaLogada)) {
            throw new ConflitoException("Nao e permitido desativar seu proprio usuario administrador.");
        }

        usuario.setAtivo(false);
        usuario.setUpdatedAt(LocalDateTime.now());
        usuarioRepository.save(usuario);
    }

    private Usuario buscarAtivoPorId(Long id) {
        Usuario usuario = usuarioRepository.findById(id)
                .orElseThrow(() -> new RecursoNaoEncontradoException("Usuario nao encontrado para o id " + id));

        if (!usuario.isAtivo()) {
            throw new RecursoNaoEncontradoException("Usuario inativo para o id " + id);
        }

        return usuario;
    }

    private UsuarioResponse toResponse(Usuario usuario) {
        return new UsuarioResponse(
                usuario.getId(),
                usuario.getMatricula(),
                usuario.getNome(),
                usuario.getCargo(),
                usuario.getRole(),
                usuario.isAtivo()
        );
    }
}

File: expedfacil-FrontEnd/src/services/usuarioService.ts
import api from "./api";

export interface UsuarioAdmin {
  id: number;
  matricula: number;
  nome: string;
  cargo: "ENCARREGADO" | "SUPERVISOR" | "GERENTE" | "CONFERENTE" | "ESTOQUISTA" | "EMPILHADEIRA";
  role: "ADMIN" | "USER";
  ativo: boolean;
}

export interface UsuarioCreateDTO {
  matricula: number;
  nome: string;
  cargo: UsuarioAdmin["cargo"];
  senha: string;
}

export interface UsuarioUpdateDTO {
  nome: string;
  cargo: UsuarioAdmin["cargo"];
  senha?: string;
}

export async function getUsuarios(page: number = 0, size: number = 10) {
  const response = await api.get(`/usuario?page=${page}&size=${size}`);
  return response.data;
}

export async function criarUsuario(payload: UsuarioCreateDTO) {
  const response = await api.post("/usuario", payload);
  return response.data;
}

export async function atualizarUsuario(id: number, payload: UsuarioUpdateDTO) {
  const response = await api.put(`/usuario/${id}`, payload);
  return response.data;
}

export async function redefinirSenha(id: number, senha: string) {
  const response = await api.patch(`/usuario/${id}/senha`, { senha });
  return response.data;
}

export async function desativarUsuario(id: number) {
  const response = await api.delete(`/usuario/${id}`);
  return response.data;
}
